src/Uno.UI/UI/Xaml/Controls/TreeView/TreeViewViewModel.cs

```diff
@@ -0,0 +1,873 @@
+			{
+				UpdateNodeSelection(selectNode, selectionState);
+
+				if (!IsInSingleSelectionMode())
+				{
+					UpdateSelectionStateOfDescendants(selectNode, selectionState);
+					UpdateSelectionStateOfAncestors(selectNode);
+				}
+			}
+		}
+
+		private void UpdateSelectionStateOfDescendants(TreeViewNode targetNode, TreeNodeSelectionState selectionState)
+		{
+			if (selectionState == TreeNodeSelectionState.PartialSelected) return;
+
+			foreach (var childNode in targetNode.Children)
+			{
+				UpdateNodeSelection(childNode, selectionState);
+				UpdateSelectionStateOfDescendants(childNode, selectionState);
+				NotifyContainerOfSelectionChange(childNode, selectionState);
+			}
+		}
+
+		private void UpdateSelectionStateOfAncestors(TreeViewNode targetNode)
+		{
+			var parentNode = targetNode.Parent;
+			if (parentNode != null)
+			{
+				// no need to update m_originalNode since it's the logical root for TreeView and not accessible to users
+				if (parentNode != m_originNode)
+				{
+					var previousState = NodeSelectionState(parentNode);
+					var selectionState = SelectionStateBasedOnChildren(parentNode);
+
+					if (previousState != selectionState)
+					{
+						UpdateNodeSelection(parentNode, selectionState);
+						NotifyContainerOfSelectionChange(parentNode, selectionState);
+						UpdateSelectionStateOfAncestors(parentNode);
+					}
+				}
+			}
+		}
+
+		private TreeNodeSelectionState SelectionStateBasedOnChildren(TreeViewNode node)
+		{
+			bool hasSelectedChildren = false;
+			bool hasUnSelectedChildren = false;
+
+			foreach (var childNode in node.Children)
+			{
+				var state = NodeSelectionState(childNode);
+				if (state == TreeNodeSelectionState.Selected)
+				{
+					hasSelectedChildren = true;
+				}
+				else if (state == TreeNodeSelectionState.UnSelected)
+				{
+					hasUnSelectedChildren = true;
+				}
+
+				if ((hasSelectedChildren && hasUnSelectedChildren) ||
+					state == TreeNodeSelectionState.PartialSelected)
+				{
+					return TreeNodeSelectionState.PartialSelected;
+				}
+			}
+
+			return hasSelectedChildren ? TreeNodeSelectionState.Selected : TreeNodeSelectionState.UnSelected;
+			throw new NotImplementedException();
+		}
+
+		private void NotifyContainerOfSelectionChange(TreeViewNode targetNode, TreeNodeSelectionState selectionState)
+		{
+			if (m_TreeViewList != null)
+			{
+				var container = m_TreeViewList.ContainerFromNode(targetNode);
+				if (container != null)
+				{
+					var targetItem = (TreeViewItem)container;
+					targetItem.UpdateSelectionVisual(selectionState);
+				}
+			}
+		}
+
+		internal IList<TreeViewNode> SelectedNodes => m_selectedNodes;
+
+		internal IList<object> SelectedItems => m_selectedItems;
+
+		internal TreeViewNode GetAssociatedNode(object item)
+		{
+			return m_itemToNodeMap[item]; //TODO: Throw or null?
+		}
+
+		internal bool IndexOfNode(TreeViewNode targetNode, out int index)
+		{
+			//return GetVectorInnerImpl().IndexOf(targetNode, index);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this commented out code.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5156819)